### PR TITLE
Async read file

### PR DESCRIPTION
**Pull request by sigmaSd**
Original discussion here https://github.com/lemunozm/termchat/pull/16

### CODE REVIEW

File: src/application.rs [hunk 1]
```diff
@@ -62,14 +67,20 @@ impl Application {
         let mut event_queue = EventQueue::new();
 
         let sender = event_queue.sender().clone(); // Collect network events
-        let network = NetworkManager::new(move |net_event| sender.send(Event::Network(net_event)));
+        let network = NetworkManager::new(move |net_event| {
+            sender.send_with_priority(Event::Network(net_event))
```
Comment (lemunozm) [inline src/application.rs]: Why did you change it to `with_priority`? Is it really needed?

File: src/application.rs [hunk 2]
```diff
@@ -62,14 +67,20 @@ impl Application {
...
+        });
 
         let sender = event_queue.sender().clone(); // Collect terminal events
         let _terminal_events = TerminalEventCollector::new(move |term_event| match term_event {
-            Ok(event) => sender.send(Event::Terminal(event)),
+            Ok(event) => sender.send_with_priority(Event::Terminal(event)),
```
Comment (lemunozm) [inline src/application.rs]: Same as before

File: src/application.rs [hunk 3]
```diff
@@ -106,6 +119,41 @@ impl Application {
...
+
+                    if let Err(e) = try_send() {
+                        // we dont have the file_name here
+                        // we'll just stop the last progress
+                        state.progress_stop_last();
+                        let msg = format!("Error sending file. error: {}", e);
+                        state.add_message(termchat_message(msg, TermchatMessageType::Error));
+                    }
+
+                    ui::draw(&mut self.terminal, &state)?;
```
Comment (lemunozm) [inline src/application.rs]: I think that you do not need this line, because it will be called at the end of the loop in line 338

**Review by lemunozm — CHANGES_REQUESTED** [review]
I really like how it looks!!

I left some minor details to review.
Also, I test it in local and `termchat` hangs from the file sender without showing the progress bar. Could you confirm that the last commit works?

File: src/application.rs (hunk 1 above)
Comment (sigmaSd) [reply src/application.rs]: Yes the sending events are too fast , without this termchat does not respond to term and network events till its finished

File: src/application.rs (hunk 3 above)
Comment (sigmaSd) [reply src/application.rs]: yes I missed that.

**Comment by sigmaSd** [comment]
Yes I can confirm it works.

If it hangs maybe somehow its picking up an old version of message-io, sometime cargo is weird and reads Cargo.toml of the parent folder(I'm saying this because that happened to me and I had to debug it)

Maybe just try with a new clean folder and make sure firewall is permissive.

File: src/application.rs (hunk 1 above)
Comment (lemunozm) [reply src/application.rs]: You are right, the `ReadFile` events saturate the queue. 

What about sending the new chunk only when the previous has been processed? In this way, the events will be dispatched without fill the `termchat` memory application, only on demand. Otherwise, if the file is huge, termchat can use all the computer memory easily.

In fact, my issue that blocks termchat while is sending a file, fills the computer memory. Maybe is related.

Comment (sigmaSd) [reply src/application.rs]: The chunks are processed immediately so I doubt that that would change anything.

Its actually weird that termchat gets filled with the file memory since its reading it in chunks, I'll have to investigate that.

Comment (lemunozm) [reply src/application.rs]: The thread that reads the file could be faster than the main thread that reads the event and sends the message?

Comment (lemunozm) [reply src/application.rs]: In that case, the event_queue is filled with chunks

Comment (sigmaSd) [reply src/application.rs]: That actually makes a lot of sense.

But how to handle this, do I have to rate limit reading somehow?

Comment (sigmaSd) [reply src/application.rs]: @lemunozm can you test this branch https://github.com/lemunozm/termchat/commit/7043901332f88f5caebbcbce6b2dbd40b7649127

This should fix the sending, if this idea work I'll work on the receiving later

Comment (sigmaSd) [reply src/application.rs]: Actually this appears to be do the trick https://github.com/sigmaSd/termchat/tree/park

I sent a 200mb file without the receiver nor the sender exceeding more than 1mb usage.

Also now this works without send_with_priority

Let me know what you think

Comment (lemunozm) [reply src/application.rs]: > But how to handle this, do I have to rate limit reading somehow?

The file should be read a new chunk once the previous chunk has been sent. 
I will see your branches during the day :) Thanks!

Comment (sigmaSd) [reply src/application.rs]: Thats what the new branch does

Comment (lemunozm) [reply src/application.rs]: Hi @sigmaSd, `termchat` continues blocking while sending the file :(

Commit: "d15fa5bc" (your park branch), in a clean state. The issue happens even if the file is not found and even if is the only termchat in the network. I have to debug it to clarify something. Maybe some thread is blocked on my computer for some unknown reason.

Regarding the approach, I think that there is no need to create more threads than the main one if once the file is open, the descriptor is stored in the application. Each new ReadFile event will send the chunk and will read more data and enqueue other ReadFile event, and so on.

Comment (lemunozm) [reply src/application.rs]: I want to add a `file-transfer` example to message-io because I think that this use case could be really useful

Comment (sigmaSd) [reply src/application.rs]: I tried to remove the threads https://github.com/sigmaSd/termchat/tree/without_threads but the problem is the event queue will be filled up with chunks again and I cant use thread::park trick anymore to throttle the reading.

Otherwise I think threads should be ok , the only problem is the issue you're having, maybe you gdb -p (pidof termchat) and use backtrace to see where its hanging.

I'll also see your message-io sending file example for more ideas.

Comment (sigmaSd) [reply src/application.rs]: Also the above method that I tried blocks everything.

I think your idea to save  file descriptors means I'll have to add at least one new thread that polls all the new files, instead of the park method that creates a thread per file sent, If that is the case, I think the advantage is not that important

Comment (sigmaSd) [reply src/application.rs]: Regarding your issue, by blocking you mean that you see the progress bar filling but you can't input keys and receive network events till its done?

If that's the case, the only thing I can think of is that std::thread::park is not being respected , it does say on the readme that the thread can wake up even with park https://doc.rust-lang.org/std/thread/fn.park.html .But that would be surprising.

I'll just for sanity push a branch that uses mutex synchronisation instead, so you can test it

Comment (sigmaSd) [reply src/application.rs]: @lemunozm here is a branch with condvar/mutex synchro which should be immune to spurious wake-ups. https://github.com/sigmaSd/termchat/tree/condvar

Comment (lemunozm) [reply src/application.rs]: Thanks for your investigation! I have finished the `file-transfer` example in branch: https://github.com/lemunozm/message-io/tree/example/file_transfer .

It works, but randomly, with long files (>100M) the receiver stops receiving. It is possible that the issue is related to some network saturation that message-io is not handling correctly (the socket is in would block state when it happens and never recovery), and is possible to be related with the `termthat` issues.

> I think your idea to save file descriptors means I'll have to add at least one new thread that polls all the new files, instead of the park method that creates a thread per file sent, If that is the case, I think the advantage is not that important

Not exactly, the branch I have upload show it :)

> Regarding your issue, by blocking you mean that you see the progress bar filling but you can't input keys and receive network events till its done?

Input blocked (ctrl-c doesn't works for example), and no command or progress bar is printed on the screen of the sender. The command is printed in the receiver and this termchat continues working.

I will test your `condvar` branch in a while, thanks!

Comment (sigmaSd) [reply src/application.rs]: Your example is clever! I implemented the idea https://github.com/sigmaSd/termchat/tree/clever_sync maybe this will work without issues.

Comment (lemunozm) [reply src/application.rs]: There is no much error handling like in termchat but the example focus on to the point :)

I was testing the message-io example issue and I pretty sure that is not related to the network/threads. It is an encoding problem. It seems to be related when a socket read data that contains more than one message at once. I thought that I was taking that into account but seems to be buggy. Randomly, the receiver is a little bit slower than the sender, and when you try to read data, there is more than a message. At that moment, the encoding header loses and the decoder thinks that the message has infinity size. Not sending any more packets.

Maybe the termchat problem is related, and my computer crashes because of computer timings or how the OS join the packets

Comment (sigmaSd) [reply src/application.rs]: Well till that's fixed I'm probably going to merge https://github.com/sigmaSd/termchat/tree/clever_sync to this pr. It looks correct and it works on my setup.

Do you have some remark regarding it?

Also do you have a test case for the issue you're having with message-io? I'm actually currently not able to reproduce the issue, but It happened sporadically before and your diagnosis aligns with what I experienced.

Comment (lemunozm) [reply src/application.rs]: Hi @sigmaSd , Yes, I made a UT that reproduce the encoding issue and also I merged the fix along with 0.4.5 version to `message-io`. Your branch should work fine with this version :)

The `clever_sync` branch looks good! 
I would write some minor comment when you merge it to this PR

Thanks!

Comment (sigmaSd) [reply src/application.rs]: Merged + updated to the new message-io, Hopefully its fixed now!

File: src/application/read_event.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        (self.callback)(file, file_name, file_size, send_id);
+        self.id += 1;
+
+        Ok(send_id)
+    }
+}
+
+use super::Event;
+pub fn read_file(
+    sender: message_io::events::EventSender<Event>,
```
Comment (lemunozm) [inline src/application/read_event.rs]: If this function returns a `Result<Chunk>` then there is no need to pass the sender as a parameter. It's the caller who knows what to do with this chunk (in this case, sending by the `EventSender`). In this way, there is no need for `read_event` module to know about the senders or events applications (and NetEvent and Event enums can remain privates in application)

File: src/application/read_event.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    mut file: std::fs::File,
```
Comment (lemunozm) [inline src/application/read_event.rs:39]: I like the idea to keep the file here to avoid store it :)

File: src/application/read_event.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    pub fn send(&mut self, file_name: String, path: std::path::PathBuf) -> Result<usize> {
+        use std::convert::TryInto;
+
+        let try_read = || -> Result<(std::fs::File, usize)> {
+            let file_size = std::fs::metadata(&path)?.len().try_into()?;
+            let file = std::fs::File::open(path)?;
+            Ok((file, file_size))
+        };
+
+        let (file, file_size) = match try_read() {
```
Comment (lemunozm) [inline src/application/read_event.rs]: Could be simply `try_read()?` ?

Comment (lemunozm) [reply src/application/read_event.rs]: Even, I think that try_read is no longer necessary

**Review by lemunozm — CHANGES_REQUESTED** [review]

**Comment by lemunozm** [comment]
The blocking problem sadly continues, I have tried to debug it and I have realized that if I comment this line: https://github.com/lemunozm/termchat/pull/17/files#diff-08f48d7f7b7769cab47740b2d52240222de76acaf454595315bb186af574a2e9R27 it works.

Does this mean that the problem is not related to the file/ event / network, maybe something when it draws / updates the progress bar? I will check it more deeply when I have time.

Note: The block happens even if the current termchat is the only termchat. Once enter key is pressed, any changes in the app and no more events comes. CPU increases to 100% and memory grow up. Seems like an infinite loop.

**Comment by sigmaSd** [comment]
Commenting that line basically removes all the progress bar drawing logic, so its very probable that there is a bug there, I'll try to find a solution.

**Comment by lemunozm** [comment]
I have found the line that blocks into an infinite loop: https://github.com/lemunozm/termchat/pull/17/files#diff-acf9ca25b7b2d13e11c1035ebbda4140466603739b73a371b67a9a267d0ce626R109

Seems like `ui_current` take a huge value

**Comment by sigmaSd** [comment]
Which line (the link is incorrect?),

Anyway nice find, this probably should be an easy fix

**Comment by lemunozm** [comment]
Sorry, the link doesn't work. It's `ui.rs:109`: https://github.com/sigmaSd/termchat/blob/read_file_event/src/ui.rs#L109

**Comment by sigmaSd** [comment]
Can you try this and see if it fix it?
```rust
// round to 2digit
let ui_current = current as f32 * ui_step;
let ui_current = ((ui_current * 100.).round() / 100.) as usize;
```

**Comment by lemunozm** [comment]
I test it, It continues blocking in the same line :(

**Comment by lemunozm** [comment]
In fact, `ui_current` is 9223372036854775808 that in binary is 2^63

**Comment by sigmaSd** [comment]
Okay the step algorithm probably fails with big numbers, I'll see what I can do

**Comment by sigmaSd** [comment]
@lemunozm 
```rust
    let width = width as f64;
    let current = current as f64;
    let max = max as f64;

    let pct = current / max;
    let pct = if !pct.is_finite() { 0.0 } else { pct };

    let ui_current = pct * width;
    let ui_remaining = width - ui_current;

    let current: String = std::iter::repeat("#").take(ui_current as usize).collect();
    let remaining: String = std::iter::repeat("-").take(ui_remaining as usize).collect();
```

taken from https://github.com/rust-lang/cargo/blob/2f115a76e5a1e5eb11cd29e95f972ed107267847/src/cargo/util/progress.rs

**Comment by lemunozm** [comment]
Thanks! I will check it. I saw fastly that when ProgressState is in Started state, the max value is 0, then this 0 is used as the denominator for the ui_step calculation. Maybe avoiding that this value is 0 can be a solution. Anyway, tomorrow I will check your changes.

**Comment by sigmaSd** [comment]
Its probably handled with is_finite() call, but yes that probably was the issue.

**Comment by sigmaSd** [comment]
I applied your suggestions + pushed the fix, hopefully now its ok!

To recap the bug (1.0 /0.0) as usize = some huge number

**Comment by lemunozm** [comment]
Now it works! 👍🏻 

I transferred successfully a 1GB of file :)
